### PR TITLE
RevisionDiff: Show BASE diff

### PR DESCRIPTION
**Pull request by gerhardol**
Part of #4154

## Proposed changes

If two revisions are selected, show the difference from the first to the BASE revision for the first-selected

This could be achieved by selecting the revisions in the grid, Navigate->Go to common ancestor (remember it), reselect revisions and finding and selecting the BASE commit again. (The last part is the most inconvenient).

Functionality is what I want, but set to WIP:
Based on #7560. 
The setting should be clarified, currently reusing AppSettings.ShowDiffForAllParents. This what I prefer, just updating the documentation.
Another reason to WIP is to get comments from @mdonatas that has implemented a solution that also includes A-BASE and B-BASE, affects a lot more code though.

image

### Before

image

### After

image

## Test methodology 

Manual

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

**Comment by mdonatas** [comment]
My solution is not prod-ready but I've been using it for several years now and rely on it every few days if not daily.

To make this more useful I've also made it so that diffs for `A-base` and `B-base` would be filtered only to files common between them. It's common that `feature` branch has a few files changed while `master` has many changed from many devs and so a diff has 40+ files and it's a hassle to either filter or scroll to find those few as you want to switch between diffs efficiently.
Continuing on the same thought I've also added another group which is same as the first (diff between `A` and `B`) but also filtered just to the files changed on both sides. I've also considered adding a diff-toolbar button to toggle "common files" but am fine with additional group for my needs.

You can try this out by building this branch https://github.com/mdonatas/gitextensions/tree/custom-build-temp it's relatively up to date to current `master`.

**Comment by gerhardol** [comment]
> p.s. I don't even want to begin to imagine what it will mean for me to rebase my changes once this gets merged

This is small change (look at the diff) that should not affect your merge, if you ever submit a PR.

I looked at your changes, but do not feel that I want to merge that. (Maybe a quick first<->selected switch is sufficient?).
But if you want to review, it will be good.

File: GitUI/UserControls/FileStatusList.cs [hunk 1]
```diff
@@ -620,11 +620,34 @@ public void SetDiffs(IReadOnlyList<GitRevision> revisions)
...
+                        // Get base commit
+                        Lazy<ObjectId> head = new Lazy<ObjectId>(() => Module.RevParse("HEAD"));
+                        var anyArtificial = parentRevs[0].IsArtificial || Revision.IsArtificial;
+                        var revA = parentRevs[0].IsArtificial ? head.Value : parentRevs[0].ObjectId;
+                        var revB = Revision.IsArtificial ? head.Value : Revision.ObjectId;
+                        ObjectId baseRevGuid = Module.GetMergeBase(revA, revB);
+                        if (baseRevGuid != null
+                            && baseRevGuid != revA
+                            && baseRevGuid != revB
+                            && (!anyArtificial || baseRevGuid != head.Value))
```
Comment (mstv) [inline GitUI/UserControls/FileStatusList.cs]: I think, the last condition can be removed because it is redundant.

**Comment by mstv** [comment]
I'd like to have both additional diffs `A - base` and `B - base`. Filtering for common files is a good idea, too. Could it be done as a `combined diff`?

**Comment by gerhardol** [comment]
> I'd like to have both additional diffs A - base and B - base. Filtering for common files is a good idea, too. Could it be done as a combined diff?

This is what @mdonatas proposes in #4154 and has implemented in 55200b3bb2ccca802e8d57ff5710ab2f28d1d3d3 mdonatas/custom-build-temp

The problem is that we have one commit selected and possibly many parents. To also have BASE->A will need to either change the structure (what is done in the commit) or to add another artificial diff that really means that "RevisionDiff should go looking somewhere else for A and B that was stored when adding the artificial commit". Also a problem with the presentation, how to show that the selected (B) commit is different.
I just want to start with the easy part, just insert another parent (BASE) for now.

The control for diff groups should be enhanced to not expand by default and to not calculate unless expanded if many parents.

File: GitUI/UserControls/FileStatusList.cs (hunk 1 above)
Comment (gerhardol) [reply GitUI/UserControls/FileStatusList.cs]: This was to handle selection of work tree/index and did not work correctly (should show HEAD as BASE). Rewrote.

**Comment by RussKie** [comment]
It feels we need to have a non-UI "differ" implementation that can take a set of revisions and returns a diff.
UI classes are crumbling under the load...

**Comment by gerhardol** [comment]
> It feels we need to have a non-UI "differ" implementation that can take a set of revisions and returns a diff.

The differ is not a problem, it is to keep track of the "selected" commit in the UI.
That is kind of what is implemented in 55200b3bb2ccca802e8d57ff5710ab2f28d1d3d3 
But that is a big change, see if I implement a proposal with another artificial commit -  that will be a hack though.

But that is a later story. I plan to merge this first step in a couple of days (or earlier, if RussKie approves).

**Comment by gerhardol** [comment]
@msftbot merge in 1 day

**Pull request merged**